tests/csapi/public_rooms_test.go

```diff
@@ -0,0 +1,188 @@
+					body := must.ParseJSON(t, res.Body)
+
+					must.MatchGJSON(
+						t,
+						body,
+						match.JSONKeyPresent("chunk"),
+						match.JSONKeyTypeEqual("chunk", gjson.JSON),
+					)
+
+					chunk := body.Get("chunk")
+					if !chunk.IsArray() {
+						t.Logf("chunk is not an array")
+						return false
+					}
+
+					// Track which rooms we've correctly found
+					foundRooms := make(map[string]bool)
+
+					// Check each room in the public rooms list
+					for _, roomData := range chunk.Array() {
+						// Verify required keys are present
+						must.MatchGJSON(
+							t,
+							roomData,
+							match.JSONKeyPresent("world_readable"),
+							match.JSONKeyPresent("guest_can_join"),
+							match.JSONKeyPresent("num_joined_members"),
+						)
+
+						canonicalAlias := roomData.Get("canonical_alias").Str
+						name := roomData.Get("name").Str
+						topic := roomData.Get("topic").Str
+						numMembers := roomData.Get("num_joined_members").Int()
+
+						// Skip rooms that aren't ours
+						if canonicalAlias == "" {
+							continue
+						}
+
+						// Find which of our rooms this matches
+						var matchedAlias string
+						for alias := range createdRooms {
+							expectedAlias := "#" + alias + ":hs1"
+							if canonicalAlias == expectedAlias {
+								matchedAlias = alias
+								break
+							}
+						}
+
+						if matchedAlias == "" {
+							continue // Not one of our rooms
+						}
+
+						roomConfig := createdRooms[matchedAlias]
+
+						// Verify member count
+						if numMembers != 1 {
+							t.Logf("Room %s has %d members, expected 1", matchedAlias, numMembers)
+							return false
+						}
+
+						// Verify name field
+						if roomConfig.name != "" {
+							if name != roomConfig.name {
+								t.Logf("Room %s has name '%s', expected '%s'", matchedAlias, name, roomConfig.name)
+								return false
+							}
+						} else {
+							if name != "" {
+								t.Logf("Room %s has unexpected name '%s', expected no name", matchedAlias, name)
+								return false
+							}
+						}
+
+						// Verify topic field
+						if roomConfig.topic != "" {
+							if topic != roomConfig.topic {
+								t.Logf("Room %s has topic '%s', expected '%s'", matchedAlias, topic, roomConfig.topic)
+								return false
+							}
+						} else {
+							if topic != "" {
+								t.Logf("Room %s has unexpected topic '%s', expected no topic", matchedAlias, topic)
+								return false
+							}
+						}
+
+						// Mark this room as correctly found
+						foundRooms[matchedAlias] = true
+						t.Logf("Successfully validated room %s", matchedAlias)
+					}
+
+					// Check if we found all our rooms
+					if len(foundRooms) != len(createdRooms) {
+						missing := []string{}
+						for alias := range createdRooms {
+							if !foundRooms[alias] {
+								missing = append(missing, alias)
+							}
+						}
+						t.Logf("Missing rooms in public list: %v (found %d/%d)", missing, len(foundRooms), len(createdRooms))
```

Comment:
This feels pretty weird and misleading as a room only gets added to `foundRooms` if it was found **and** it has all of the expected fields.

I would refactor away from this pattern.